services/storage/tests/unit/test_handlers_files.py

```diff
@@ -873,6 +873,128 @@ async def test_download_file(
     assert filecmp.cmp(uploaded_file, dest_file)
 
 
+@pytest.mark.parametrize(
+    "file_size",
+    [
+        pytest.param(parse_obj_as(ByteSize, "1")),
+    ],
+    ids=byte_size_ids,
+)
+async def test_download_file_cases(
+    client: TestClient,
+    file_size: ByteSize,
+    upload_file: Callable[[ByteSize, str], Awaitable[tuple[Path, SimcoreS3FileID]]],
+    location_id: int,
+    project_id: ProjectID,
+    node_id: NodeID,
+    user_id: UserID,
+    create_empty_directory: Callable[..., Awaitable[FileUploadSchema]],
+    create_file_of_size: Callable[[ByteSize, str | None], Path],
+    storage_s3_client: StorageS3Client,
+    storage_s3_bucket: S3BucketName,
+):
+    assert client.app
+
+    # 1. error case
+    # no file was not uploaded
+    missing_file = parse_obj_as(SimcoreS3FileID, f"{project_id}/{node_id}/missing.file")
+    assert (
+        await storage_s3_client.file_exists(storage_s3_bucket, s3_object=missing_file)
+        is False
+    )
+
+    download_url = (
+        client.app.router["download_file"]
+        .url_for(
+            location_id=f"{location_id}",
+            file_id=urllib.parse.quote(missing_file, safe=""),
+        )
+        .with_query(user_id=user_id)
+    )
+    response = await client.get(f"{download_url}")
+    data, error = await assert_status(response, web.HTTPNotFound)
+    assert data is None
+    assert missing_file in error["message"]
+
+    # 2. file_meta_data entry corresponds to a file
+    # upload a single file as a file_meta_data entry and check link
+    _, uploaded_file_uuid = await upload_file(file_size, "meta_data_entry_is_file.file")
+    assert (
+        await storage_s3_client.file_exists(
+            storage_s3_bucket, s3_object=uploaded_file_uuid
+        )
+        is True
+    )
+
+    download_url = (
+        client.app.router["download_file"]
+        .url_for(
+            location_id=f"{location_id}",
+            file_id=urllib.parse.quote(uploaded_file_uuid, safe=""),
+        )
+        .with_query(user_id=user_id)
+    )
+    response = await client.get(f"{download_url}")
+    data, error = await assert_status(response, web.HTTPOk)
+    assert not error
+    assert data
+    assert "link" in data
+    assert parse_obj_as(AnyHttpUrl, data["link"])
+
+    # 3. file_meta_data entry corresponds to a directory
+    # upload a file inside a directory and check the download link
+
+    directory_name = "a-test-dir"
+    directory_file_upload = await create_empty_directory(directory_name)
```

Comment:
why not using temp_path like above? this looks like you are re-inventing built-in fixtures.

Reply:
I don't understand this comment. But this fixture is used elsewhere and it crates an empty directory. It also uses temp_path.